yeat/cli/config.py

```diff
@@ -8,44 +8,46 @@
 # -------------------------------------------------------------------------------------------------
 
 import json
+from pathlib import Path
 from warnings import warn
 
 
-ALGORITHMS = ("spades", "megahit", "unicycler", "flye", "canu")
-KEYS = ["algorithm", "extra_args"]
+KEYS1 = ("samples", "assemblers")
+KEYS2 = ("label", "algorithm", "extra_args", "samples")
+PAIRED = ("spades", "megahit", "unicycler")
+PACBIO = ("canu", "flye")
+ALGORITHMS = PAIRED + PACBIO
 
 
 class AssemblyConfigurationError(ValueError):
     pass
 
 
 class AssemblerConfig:
-    def __init__(self, algorithm, extra_args=""):
-        if algorithm not in ALGORITHMS:
-            raise AssemblyConfigurationError(f"Unsupported assembly algorithm '{algorithm}'")
-        self.algorithm = algorithm
-        self.extra_args = extra_args
+    def __init__(self, data, threads):
+        self.data = data
+        self.validate()
+        self.threads = threads
+        self.create_sample_and_assembler_objects()
+        self.batch()
 
     @classmethod
-    def from_json(cls, data):
-        return cls(data["algorithm"], data["extra_args"])
-
-    @staticmethod
-    def parse_json(instream):
-        configdata = json.load(instream)
-        for entry in configdata:
-            AssemblerConfig.validate(entry)
-        algorithms = [entry["algorithm"] for entry in configdata]
-        if len(algorithms) > len(set(algorithms)):
-            message = "Duplicate assembly configuration: please check config file"
+    def from_json(cls, infile, threads):
+        data = json.load(open(infile))
+        return cls(data, threads)
+
+    def validate(self):
+        self.check_keys(self.data, KEYS1)
+        for assembler in self.data["assemblers"]:
+            self.check_keys(assembler, KEYS2)
```

Comment:
The `KEYS1` and `KEYS2` variables are defined as constants at the top of the file but are used in only one place, for validation. I'd suggest it would be clearer to just declare them in-line in this code.

```python
        self.check_keys(self.data, ("samples", "assemblers"))
        for assembler in self.data["assemblers"]:
            self.check_keys(assembler, ("label", "algorithm", "extra_args", "samples"))
```